align/primitive/default/canvas.py

```diff
@@ -71,6 +71,12 @@ def _create_metal( self, layer, info, *, check=True):
                     spg_pitch, spg_stop, spg_offset = (nm_pitch,
                                                        self._get_via_ext(base_layer, nv),
                                                        self._get_metal_offset(nm))
+
+                # Stoppoint shall satisfy both upper and lower layers
+                spg_stop = max(self._get_via_ext(base_layer, nv), self._get_via_ext(base_layer, pv)) 
```

Comment:
Not sure how this helps. The spot point grid is defined with respect to one orthogonal layer. This condition helps if and only if the two layers above and below it are completely aligned (basically spg_pitch and spg_offset are exactly the same both above and below). Can you write a test for the issue you are trying to fix?

Reply:
When the extension needed for the via above and via below are different (eg., v1-m2-v2 where v2 requires a longer extension), the stoppoint grid is set according to v1. This causes DRC violations where m2 doesn't enclose v2.

Reply:
In the example you have given (v1-m2-v2)... This fix works great since M1 pitch == M3 pitch and M1 offset == M3 offset. If that's not the case, then the stoppoint pitch itself is tied to a single layer definition (either M1 or M3). The fix may still work but will draw unnecessarily long metals for the smaller metal enclosure as well.

I am more concerned about why you saw this DRC error in the first place. Handling non-aligned pitches & enclosures is the whole purpose behind snapToLegal in addWireAndMultiViaSet. Can we set up a meeting and talk through a counterexample? There may be a bug that needs to be addressed.